bridge-fungible-token/src/bridge_fungible_token_contract.sw.sw

```diff
@@ -0,0 +1,164 @@
+    parse_message_data,
+    register_refund,
+    send_message,
+    transfer_tokens,
+};
+
+////////////////////////////////////////
+// Storage declarations
+////////////////////////////////////////
+storage {
+    refund_amounts: StorageMap<(EvmAddress, EvmAddress), U256> = StorageMap {},
+}
+
+////////////////////////////////////////
+// ABI Implementations
+////////////////////////////////////////
+// Implement the process_message function required to be a message receiver
+impl MessageReceiver for Contract {
+    #[storage(read, write)]
+    fn process_message(msg_idx: u8) {
+        let input_sender = input_message_sender(1);
+
+        require(input_sender.value == LAYER_1_ERC20_GATEWAY, BridgeFungibleTokenError::UnauthorizedUser);
+
+        let message_data = parse_message_data(msg_idx);
+
+        require(message_data.l1_asset == ~EvmAddress::from(LAYER_1_TOKEN), BridgeFungibleTokenError::IncorrectAssetDeposited);
+
+        if message_data.l1_asset != ~EvmAddress::from(LAYER_1_TOKEN)
+        {
+            // Register a refund if tokens don't match. The L1 tokens are now locked in the contract on Ethereum, so reverting here is not the correct action as it would prevent the registration of a claimable refund on the Fuel side of the bridge.
+            register_refund(message_data.from, message_data.l1_asset, amount);
+        } else {
+            // The value needs to be converted from the Ethereum side decimals (18) into the Fuel side decimals (9).
+            // This could result in a refund (value too large to fit in u64 under new decimals or too small to fit in new decimals)
+            /**
+                So, "1" erc20 token is represented internally (in ethereum contract) as 1 * 10^18 or 1_000_000_000_000_000_000
+                Dividing this number by 10^9 (because of 9 decimals in fuel contract) results in 1_000_000_000 internally, which is "1" token in the fuel contract
+                Internal Representations:
+                Ethereum: 1 = 0.000_000_000_000_000_001 ether (1 wei)
+                Fuel:     1 = 0.000_000_001 Base Asset
+                so: 999_999_999 wei sent from ethereum to fuel would result in a refund because it's too small to fit in the 9 decimals observed in the L2 BridgeFungibleToken contract, because this would be
+                              `0.000_000_000_999_999_999`, when
+                              `0.000_000_001` is the smallest value we can work with at 9 decimals.
+            */
+
+            let decomposed = decompose(message_data.amount);
+            let amount = ~U256::from(decomposed.0, decomposed.1, decomposed.2, decomposed.3);
+            let l1_amount_opt = amount.as_u64();
+            match l1_amount_opt {
+                Result::Err(e) => {
+                    register_refund(message_data.from, message_data.l1_asset, amount);
+                },
+                Result::Ok(amount) => {
+                    mint_tokens(amount, Identity::Address(message_data.to));
```

Comment:
I'm not sure we want to be using `Identity` when minting tokens. Currently, all bridged assets must be going to an `Address` and cannot be a `ContractId`. We need to eventually think about how to best support depositing directly to a contract in the future.

Reply:
I've now switched this to only allow minting to `address`es.
Agreed that we'll need to consider how to best support depositing directly to a contract in a future iteration.